drivers/Aqara/aqara-lock/src/credential_utils.lua

```diff
@@ -0,0 +1,134 @@
+
+  device:set_field(HOST_COUNT, host_cnt, { persist = true })
+  credential_utils.save_data(driver)
+end
+
+credential_utils.sync_all_credential_info = function(driver, device, command)
+  for _, credentialinfo in ipairs(command.args.credentialInfo) do
+    if credentialinfo.userType == "host" then
+      credential_utils.update_remote_control_status(driver, device, true)
+    end
+  end
+  device:emit_event(lockCredentialInfo.credentialInfo(utils.deep_copy(command.args.credentialInfo),
+    { visibility = { displayed = false } }))
+  credential_utils.save_data(driver)
+end
+
+credential_utils.upsert_credential_info = function(driver, device, command)
+  if #command.args.credentialInfo == 0 then
+    return
+  end
+
+  local credentialInfoTable = utils.deep_copy(device:get_latest_state("main", lockCredentialInfo.ID,
+    lockCredentialInfo.credentialInfo.NAME, {}))
+
+  if #credentialInfoTable > 0 then
+    credentialInfoTable = utils.update(credentialInfoTable, command.args.credentialInfo);
+  end
+
+  if credential_utils.is_exist_host(device) == false then
+    credential_utils.update_remote_control_status(driver, device, true)
+  end
+
+  credentialInfoTable = utils.merge(credentialInfoTable, command.args.credentialInfo)
+
+  device:emit_event(lockCredentialInfo.credentialInfo(utils.deep_copy(credentialInfoTable),
+    { visibility = { displayed = false } }))
+  credential_utils.save_data(driver)
+end
+
+credential_utils.delete_user = function(driver, device, command)
+  local credentialInfoTable = utils.deep_copy(device:get_latest_state("main", lockCredentialInfo.ID,
+    lockCredentialInfo.credentialInfo.NAME, {}))
+  if #credentialInfoTable == 0 then
+    return
+  end
+
+  local id = tostring(command.args.userId)
+
+  for i = #credentialInfoTable, 1, -1 do
+    if id == credentialInfoTable[i].userId then
+      if credentialInfoTable[i].userType == "host" then
+        credential_utils.update_remote_control_status(driver, device, false)
+      end
+      table.remove(credentialInfoTable, i)
+    end
+  end
+
+  device:emit_event(lockCredentialInfo.credentialInfo(utils.deep_copy(credentialInfoTable),
+    { visibility = { displayed = false } }))
+  credential_utils.save_data(driver)
+end
+
+credential_utils.delete_credential = function(driver, device, command)
+  local credentialInfoTable = utils.deep_copy(device:get_latest_state("main", lockCredentialInfo.ID,
+    lockCredentialInfo.credentialInfo.NAME, {}))
+  if #credentialInfoTable == 0 then
+    return
+  end
+
+  for i = #credentialInfoTable, 1, -1 do
+    if command.args.credentialId == credentialInfoTable[i].credentialId then
+      if credentialInfoTable[i].userType == "host" then
+        credential_utils.update_remote_control_status(driver, device, false)
+      end
+      table.remove(credentialInfoTable, i)
+      break
+    end
+  end
+
+  device:emit_event(lockCredentialInfo.credentialInfo(utils.deep_copy(credentialInfoTable),
+    { visibility = { displayed = false } }))
```

Comment:
I don't know if I saw an answer here, but is there a reason you have to use `deep_copy` here?

Reply:
Did you mean removing all the "deep_copy" in "emit_event"? I misunderstood that I should just remove the line. Let me check it.

Reply:
just removing the `deep_copy`. I think 

```
  device:emit_event(lockCredentialInfo.credentialInfo(credentialInfoTable,
    { visibility = { displayed = false } }))
```

should work, right?

Reply:
Yeah, you're right